### PR TITLE
Ksagiyam/fix things around bc

### PR DESCRIPTION
**Pull request by ksagiyam**
This PR is to fix two things:

1.
```
mesh = UnitSquareMesh(1, 1)
V = VectorFunctionSpace(mesh, "CG", 1)
g = as_vector([0., 0.])
bc = DirichletBC(V, g, "on_boundary")

print(repr(g))  # Zero((2,), (), ())
print(repr(bc.function_arg))  # currently prints Zero((), (), ()) <- fix
```

2.
```
mesh = UnitSquareMesh(1, 1)
V = FunctionSpace(mesh, "CG", 1)
W0 = V * V
W1 = V * V * V * V

print(W0 == W1)  # currently prints True <- fix
```

### CODE REVIEW

**Comment by wence-** [comment]
Can you add a test that exercises the failing example from #2396

**Comment by ksagiyam** [comment]
Good idea. Added!

File: tests/regression/test_matrix_free.py [hunk 1]
```diff
@@ -308,3 +308,34 @@ def test_duplicate(a, bcs):
         B_petsc.mult(x, y)
     # Check if original rhs is equal to BA^-1 (rhs)
     assert np.allclose(rhs.vector().array(), solution2.vector().array())
+
+
+def test_matrix_free_fieldsplit_with_real():
```
Comment (dham) [inline tests/regression/test_matrix_free.py:313]: Is this really matrix free? I can't see anything in the solver options that would make it so.

**Review by dham — CHANGES_REQUESTED** [review]

File: tests/regression/test_matrix_free.py (hunk 1 above)
Comment (ksagiyam) [reply tests/regression/test_matrix_free.py:313]: Turned out that, when we have `Real`, `solving_utils.set_defaults()` puts `matfree` here: https://github.com/firedrakeproject/firedrake/blob/06125881584cc898fed1f9b9c37a25bbe3bb55c9/firedrake/solving_utils.py#L95

Comment (connorjward) [reply tests/regression/test_matrix_free.py:313]: Perhaps do an assertion for `stokes_solver.parameters["mat_type"]` to make this explicit?

Comment (ksagiyam) [reply tests/regression/test_matrix_free.py:313]: Or can I just add comments, as `set_defaults()` also sets many related `fieldsplit` options?

Comment (ksagiyam) [reply tests/regression/test_matrix_free.py:314]: Actually, I agree. It's probably better to spell out options we need here; updated.

**Review by dham — APPROVED** [review]

**Pull request merged**